### ISSUE TITLE
Custom entry regression (?) from 8.10 to 8.12?

### ISSUE THREAD

**Issue by bcpierce00** [coq/coq#12455]
The following example compiles in both 8.10 and 8.12, but uncommenting the next-to-last command causes it to fail on 8.12 only.  Is this expected behavior? If so, can someone tell me how to achieve the effect of "application syntax" in a custom grammar?  Thank you!

```
Declare Custom Entry com.
Notation "[[ e ]]" := e (e custom com at level 0).
Notation "x" := x (in custom com at level 0, x constr at level 0).

(* UNCOMMENTING THIS CAUSES THE NEXT LINE TO FAIL IN 8.12
Notation "f x" := (f x)
                  (in custom com at level 0,
                  f constr, x constr, only parsing).
*)

Definition example := [[ 3 ]].
```

**Comment by ejgallego** [comment]
Let's research this before we release 8.12, @herbelin any immediate idea?

**Comment by herbelin** [comment]
This is indeed a change of behavior but of the "bug fix" kind, not of the "regression" style. Indeed, 8.10 had a "severe" bug which was causing
```
Notation "f x" := (f x)
                  (in custom com at level 0,
                  f constr, x constr, only parsing).
```
to be wrongly interpreted as:
```
Notation "f x" := (f x)
                  (in custom com at level 0,
                  f custom com, x custom com, only parsing).
```
If you want to preserve the 8.10 behavior (but using the right incantation for it!), use the latter (with `com`). If you really intend to parse any application of a `constr` to a `constr`, you can do it, typically with:
```
Notation "f x" := (f x)
                  (in custom com at level 0,
                  f constr at level 0, x constr at level 9, only parsing).
```
so that it factorizes with the `constr at level 0` of the `"x"` rule (while the `x` at level `9` is to ensure that `x` excludes the level 10, where applications are parsed).

Incidentally, @bcpierce00, I'd be interested in more feedback about custom notations. For instance, there is already an ongoing extension to numerals #11561 but I'd be more generally interested in other use cases of the custom mechanism.

**Comment by bcpierce00** [comment]
Many thanks, Hugo, for the quick response!  Your last suggestion for the better way to parse applications seems to work. (This fragment was extracted from our syntax for Hoare Logic assertions in SF, where we want to be able to parse things like [[ Y = 0 /\ Z = factorial X ]] without heavy brackets to escape from the assertion syntax back to the main grammar.)

Something else failed a few lines below, though. I’ll investigate a little more and submit a new issue if needed.
> Incidentally, @bcpierce00 <https://github.com/bcpierce00>, I'd be interested in more feedback about custom notations. For instance, there is already an ongoing extension to numerals #11561 <https://github.com/coq/coq/pull/11561> but I'd be more generally interested in other use cases of the custom mechanism.
> 
Last fall we used custom notations very heavily in a major revision of the first two volumes of SF, but at the end there were a number of questions and points that we were not satisfied with; we had intended to clean things up this semester, but external events in the world slowed us down.  We are just getting started doing this work in preparation for a new release. If you have a few minutes, we would love to have your feedback on whether we are doing things in the best way.  

We can extract a more complete version of our definitions and a few longer examples to a separate file; alternatively, I can give you direct access to the SF repo so that you can see everything “in situ” (which will allow you to inspect all the places where our notations are used).

**Comment by bcpierce00** [comment]
P.S.  Although the example now compiles, I am not sure I understand why: there is no instance of "something followed by something else" in the last line, so why does the production for parsing such things matter?

**Comment by herbelin** [comment]
Yes, I'd be interested in knowing more about the situations which you were not satisfied with, and to understand if it is inherent to the custom notations, to camlp5, or to some other issues, ideally in an interactive way.

> P.S. Although the example now compiles, I am not sure I understand why: there is no instance of "something followed by something else" in the last line, so why does the production for parsing such things matter?

If the question is why adding a rule to parse `[[ t u ]]` (for `t` and `u` constr) breaks the parsing of `[[ 3 ]]`, it is a limitation of camlp5 which does not compile rules to its internal LL automaton but which instead expects the user to give the rules already in LL-compiled form. The grammar, as shown by `Print Custom Grammar com` is:
```
[ "0" RIGHTA
  [ constr:operconstr LEVEL "200"; constr:operconstr LEVEL "200"
  | constr:operconstr LEVEL "0" ] ]
```
Once camlp5 enters `constr:operconstr LEVEL "200"` (in which it finds `3` as a particular case), it expects to find another `constr:operconstr LEVEL "200"`  which it does not find. Otherwise said, there is no backtracking across rules once one of them started to consume the input.

Maybe more illustrating is to swap the order of declaration of the `"x"` rule and the `"f x"` rule, so as to obtain:
```
[ "0" RIGHTA
  [ constr:operconstr LEVEL "0"
  | constr:operconstr LEVEL "200"; constr:operconstr LEVEL "200" ] ]
```
In this case, it'll parse `[[ 3 ]]` but not, say `[[ S 3 ]]``.

**Comment by bcpierce00** [comment]
P.S.  Thank you for the explanation of the mis-parse — that was very illuminating!  I wonder if something like this belongs in the manual…?
